src/c2s/mongoose_c2s.erl

```diff
@@ -0,0 +1,920 @@
+                  element => El, from_jid => ServerJid, to_jid => ServerJid},
+    Acc0 = mongoose_acc:new(AccParams),
+    Params = hook_arg(StateData, C2SState),
+    {_, Acc1} = mongoose_c2s_hooks:user_send_xmlel(HostType, Acc0, Params),
+    handle_state_after_packet(StateData, C2SState, Acc1).
+
+-spec handle_c2s_packet(c2s_data(), c2s_state(), exml:element()) -> fsm_res().
+handle_c2s_packet(StateData = #c2s_data{host_type = HostType}, C2SState, El) ->
+    Acc0 = element_to_origin_accum(StateData, El),
+    Acc1 = mongoose_hooks:c2s_preprocessing_hook(HostType, Acc0, StateData),
+    case mongoose_acc:get(hook, result, undefined, Acc1) of
+        drop -> {next_state, session_established, StateData};
+        _ -> do_handle_c2s_packet(StateData, C2SState, Acc1)
+    end.
+
+-spec do_handle_c2s_packet(c2s_data(), c2s_state(), mongoose_acc:t()) -> fsm_res().
+do_handle_c2s_packet(StateData = #c2s_data{host_type = HostType}, C2SState, Acc) ->
+    case mongoose_c2s_hooks:user_send_packet(HostType, Acc, hook_arg(StateData)) of
+        {ok, Acc1} ->
+            Acc2 = handle_outgoing_stanza(StateData, Acc1, mongoose_acc:stanza_name(Acc1)),
+            handle_state_after_packet(StateData, C2SState, Acc2);
+        {stop, Acc1} ->
+            handle_state_after_packet(StateData, C2SState, Acc1)
+    end.
+
+%% @doc Process packets sent by the user (coming from user on c2s XMPP connection)
+-spec handle_outgoing_stanza(c2s_data(), mongoose_acc:t(), binary()) -> mongoose_acc:t().
+handle_outgoing_stanza(StateData = #c2s_data{host_type = HostType}, Acc, <<"message">>) ->
+    TS0 = mongoose_acc:timestamp(Acc),
+    Acc1 = mongoose_c2s_hooks:user_send_message(HostType, Acc, hook_arg(StateData)),
+    Acc2 = maybe_route(Acc1),
+    TS1 = erlang:system_time(microsecond),
+    mongoose_metrics:update(HostType, [data, xmpp, sent, message, processing_time], (TS1 - TS0)),
+    Acc2;
+handle_outgoing_stanza(StateData = #c2s_data{host_type = HostType}, Acc, <<"iq">>) ->
+    Acc1 = mongoose_c2s_hooks:user_send_iq(HostType, Acc, hook_arg(StateData)),
+    maybe_route(Acc1);
+handle_outgoing_stanza(StateData = #c2s_data{host_type = HostType}, Acc, <<"presence">>) ->
+    {_, Acc1} = mongoose_c2s_hooks:user_send_presence(HostType, Acc, hook_arg(StateData)),
+    Acc1;
+handle_outgoing_stanza(StateData = #c2s_data{host_type = HostType}, Acc, _) ->
+    {_, Acc1} = mongoose_c2s_hooks:user_send_xmlel(HostType, Acc, hook_arg(StateData)),
+    Acc1.
+
+-spec handle_incoming_stanza(c2s_data(), c2s_state(), mongoose_acc:t()) -> fsm_res().
+handle_incoming_stanza(StateData = #c2s_data{host_type = HostType}, C2SState, Acc) ->
+    {From, To, El} = mongoose_acc:packet(Acc),
+    FinalEl = jlib:replace_from_to(From, To, El),
+    ParamsAcc = #{from_jid => From, to_jid => To, element => FinalEl},
+    Acc1 = mongoose_acc:update_stanza(ParamsAcc, Acc),
+    case mongoose_c2s_hooks:user_receive_packet(HostType, Acc1, hook_arg(StateData)) of
```

Comment:
* `user_receive_packet` naming is kept for backwards compatibility, it is the name it was always used. The others I could rename to keep the same convention.
* The idea is that in reality almost every handler we build only cares for `message`s, like MAM, inbox, or carbon copies, so we run handlers for iqs and presences and waste cpu time in them, and we write handlers having to remember to ignore without crashing on the other types of stanzas. It's cleaner, and faster, to subscribe exactly to what you need.

Reply:
I agree that it might be cleaner not to have to pattern match for the specific stanza types in handlers. I'm just not convinced that it's faster, because for each message we need to run two hooks instead of one, and calling a hook takes some time, including reading the ETS table. Anyway, it's good enough for me.